core/src/main/java/org/apache/calcite/plan/volcano/RelSet.java

```diff
@@ -158,11 +160,66 @@ RelSubset getOrCreateSubset(
       final VolcanoPlanner planner =
           (VolcanoPlanner) cluster.getPlanner();
 
+      // Converters from newly introduced subset to all the remaining one, only if possible to
+      // convert.  No point adding converters if it is not possible.
+      for (RelSubset other : subsets) {
+        if (other.getConvention() == Convention.NONE) {
+          continue;
+        }
+
+        assert other.getTraitSet().size() == subset.getTraitSet().size();
+
+        final ImmutableList<RelTrait> otherTrait =
+                other.getTraitSet().difference(subset.getTraitSet());
+        final ImmutableList<RelTrait> thisTrait =
+                subset.getTraitSet().difference(other.getTraitSet());
+        assert otherTrait.size() == thisTrait.size();
+
+        if (otherTrait.size() == 1 && thisTrait.size() == 1) {
+          RelTraitDef traitDef = otherTrait.get(0).getTraitDef();
+          boolean canConvert = traitDef.canConvert(
+                  cluster.getPlanner(), thisTrait.get(0), otherTrait.get(0));
+
+          if (canConvert) {
+            final AbstractConverter converter =
+                    new AbstractConverter(cluster, subset, traitDef, other.getTraitSet());
+            planner.register(converter, other);
+          }
+        }
+      }
+
       subsets.add(subset);
 
-      if (planner.root != null
-          && planner.root.set == this) {
-        planner.ensureRootConverters();
+      // Converters from newly introduced subset to all the remaining one, only if possible to
+      // convert.  No point adding converters if it is not possible.
+      for (RelSubset other : subsets) {
+        if (other == subset) {
+          continue;
+        }
+
+        if (other.getConvention() == Convention.NONE) {
```

Comment:
This doesn't seem necessary since the ConventionTraitDef finds only existing ConverterRule paths and thus should fail canConvert below. Note that this wasn't previously the case so with that fix, this can probably be removed.